src/soroban/assembled_transaction.ts

```diff
@@ -0,0 +1,717 @@
+    })
+      .addOperation(contract.call(options.method, ...(options.args ?? [])))
+      .setTimeout(TimeoutInfinite)
+      .build();
+
+    return await tx.simulate();
+  }
+
+  simulate = async (): Promise<this> => {
+    if (!this.raw) throw new Error("Transaction has not yet been assembled");
+    this.simulation = await this.server.simulateTransaction(this.raw);
+
+    if (SorobanRpc.Api.isSimulationSuccess(this.simulation)) {
+      this.raw = SorobanRpc.assembleTransaction(
+        this.raw,
+        this.simulation
+      ).build();
+    }
+
+    return this;
+  };
+
+  get simulationData(): {
+    result: SorobanRpc.Api.SimulateHostFunctionResult;
+    transactionData: xdr.SorobanTransactionData;
+  } {
+    if (this.simulationResult && this.simulationTransactionData) {
+      return {
+        result: this.simulationResult,
+        transactionData: this.simulationTransactionData,
+      };
+    }
+    // else, we know we just did the simulation on this machine
+    const simulation = this.simulation!;
+    if (SorobanRpc.Api.isSimulationError(simulation)) {
+      throw new Error(`Transaction simulation failed: "${simulation.error}"`);
+    }
+
+    if (SorobanRpc.Api.isSimulationRestore(simulation)) {
+      throw new AssembledTransaction.ExpiredStateError(
+        `You need to restore some contract state before you can invoke this method. ${JSON.stringify(
+          simulation,
+          null,
+          2
+        )}`
+      );
+    }
+
+    if (!simulation.result) {
+      throw new Error(
+        `Expected an invocation simulation, but got no 'result' field. Simulation: ${JSON.stringify(
+          simulation,
+          null,
+          2
+        )}`
+      );
+    }
+
+    // add to object for serialization & deserialization
+    this.simulationResult = simulation.result;
+    this.simulationTransactionData = simulation.transactionData.build();
+
+    return {
+      result: this.simulationResult,
+      transactionData: this.simulationTransactionData!,
+    };
+  }
+
+  get result(): T {
+    try {
+      return this.options.parseResultXdr(this.simulationData.result.retval);
+    } catch (e) {
+      if (!implementsToString(e)) throw e;
+      let err = this.parseError(e.toString());
+      if (err) return err as T;
+      throw e;
+    }
+  }
+
+  parseError(errorMessage: string): Result<never, ErrorMessage> | undefined {
+    if (!this.options.errorTypes) return undefined;
+    const match = errorMessage.match(contractErrorPattern);
+    if (!match) return undefined;
+    let i = parseInt(match[1], 10);
+    let err = this.options.errorTypes[i];
+    if (!err) return undefined;
+    return new AssembledTransaction.Result.Err(err);
+  }
+
+  getPublicKey = async (): Promise<string | undefined> => {
+    const wallet = this.options.wallet;
+    if (!(await wallet.isConnected()) || !(await wallet.isAllowed())) {
+      return undefined;
+    }
+    return (await wallet.getUserInfo()).publicKey;
+  };
```

Comment:
This doesn't feel appropriate for the SDK: I would expect this to just return the source account on the transaction. Are these these methods Freighter-specific? I don't see the interface for what `this.options.wallet` is supposed to be.

Reply:
You can see the `interface` definition here https://github.com/stellar/js-stellar-sdk/pull/891/files#diff-d7db6307c66b5050519b2117d1ba8d0bbe3a743c92b463125f0315b713d4dccaR6-R24

Yes, it's Freighter-specific for now. Which is not a new problem. This code is mostly copied from the code that currently lives in the CLI repo, which is a bad place for it. Putting it here makes the problem better by exposing the problem and allowing us to fix it in a central place, within a JS repo instead of a Rust repo. We should be using whatever the Wallet standards team is coming up with.

The biggest change here from what's currently generated in the bindings is that you are now _required_ to inject `wallet`, rather than defaulting to Freighter! A good improvement, I'm sure you'll agree! You can see that the current logic [auto-imports Freighter](https://github.com/stellar/soroban-tools/blob/deb1f7cbdc4113dbedffe981fad0dff3d5945d5c/cmd/crates/soroban-spec-typescript/src/project_template/src/assembled-tx.ts#L236-L238) if no wallet is injected.